pulp_container/app/viewsets.py

```diff
@@ -560,6 +564,63 @@ def __init__(self, blob, path, status, request, send_body=False):
         super().__init__(headers=headers, status=status)
 
 
+class ContainerRegistryApiMixin:
+    """
+    Mixin to add docker registry specific headers to all error responses.
+    """
+
+    def handle_exception(self, exc):
+        """
+        Add docker registry specific headers to all error responses.
+        """
+        response = super().handle_exception(exc)
+        response["Docker-Distribution-API-Version"] = "registry/2.0"
+        return response
+
+    def get_drv_pull(self, path):
+        """
+        Get distribution, repository and repository_version for pull access.
+        """
+        distribution = get_object_or_404(models.ContainerDistribution, base_path=path)
+        if distribution.repository:
+            repository_version = distribution.repository.latest_version()
+        elif distribution.repository_version:
+            repository_version = distribution.repository_version
+        else:
+            raise Http404("Repository {} does not exist.".format(path))
+        return distribution, distribution.repository, repository_version
+
+    def get_dr_push(self, request, path, create=False):
+        """
+        Get distribution and repository for push access.
+
+        Optionally create them if not found.
+        """
+        dist_created = False
+        if create:
+            # TODO: Use serializer to create objects
+            distribution, dist_created = models.ContainerDistribution.objects.get_or_create(
+                base_path=path, defaults={"name": path},
+            )
+        else:
+            distribution = get_object_or_404(models.ContainerDistribution, base_path=path)
+        if dist_created:
+            repository = models.ContainerPushRepository.objects.create(name=path)
```

Comment:
can we add here try/except in case there is already a ContainerRepository with such name?

Reply:
also maybe we should delete the distribution if we fail with repo creation

Reply:
Sure. What should be in the except block? Is it a 404, 409 or 500?
What if there is already a distribution, but no repository. Is auto creating the latter ok? Or does it mean someone reserved that name for later use?

Reply:
We should put repo and distribution creation into a database transaction.

Reply:
and repo_version

Reply:
as we discussed per call - if existing distribution has no repo/repo_version assigned to it --> fail whole transaction